gui: Fix incorrect UI language auto detection (fixes #9668)

tomasz1986 opened the pull request:
gui: Fix incorrect UI language auto detection (fixes #9668)

Currently, the code only checks whether the detected language partially
matches one of the available languages. This means that if the detected
language is "fi" but among the available languages there is only "fil"
and no "fi", then it will match "fi" with "fil", even though the two are
completely different languages.

With this change, the matching is only done when there is a hyphen in
the language code, e.g. "en" will match with "en-US".

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

acolomb reviewed (commented):
Wouldn't it be easier to just strip off anything following a `-` character, and then doing an exact comparison?

tomasz1986 commented:
> Wouldn't it be easier to just strip off anything following a `-` character, and then doing an exact comparison?

Done, I think so.

acolomb commented on gui/default/syncthing/core/localeService.js:
```diff
@@ -78,14 +79,16 @@ angular.module('syncthing.core')
+                                // Match "en" with "en-US" but not "fi" with "fil",
+                                // and match "en-US" with "en" but not "fil" with "fi".
+                                if (possibleLang !== lang) {
+                                    lang = lang.replace(pattern, '');
```
This could be done once, where `lang` is first assigned in the loop.

acolomb replied:
In line 71, where the value is assigned:

    lang = langs[i].replace(pattern, '');

acolomb replied:
Ah, sorry I think I see what you mean.  Simply discarding the region info would match the wrong variant when there are several.  So it's OK to do it only when not finding an exact complete match.

tomasz1986 replied:
I'm not sure the current version is correct. It may be too simplified.

There are five possible cases.

1. The detected and available languages are the same (e.g. `zh-TW` and `zh-TW`). Use the language.
2. The detected and available language are completely different with no hyphen. Use the default locale.
3. The detected language has a hyphen but the available one does not (e.g. `zh-TW` and `zh`). Use `zh`.
4. The detected language has no hyphen but the available one does (e.g. `zh` and `zh-TW`). Use `zh-TW`.
5. Both detected and available languages have hyphens but are not the same (e.g. `zh-TW` and `zh-CN`). Use `zh-CN`.

I'm testing different possibilities right now, and for example, if we set the browser to `en-GB`, the UI actually defaults to `en-US` (or rather just `en`, which translates to `en-US`), which is wrong. The same goes for `zh-TW` which matches with `zh-CN` instead of `zh-TW`. When using just `return lang === possibleLang` with the conditional above commented out, the GUI does default to `zh-TW` instead of `zh-CN`.

English seems to be a special case, because the array looks like this:

```
[
    "en",
    "en-GB"
]
```

and the language should actually match with the second one, not just `en`.

I'm going to play around with this a little bit more.

acolomb replied:
So it depends on the order of the available languages array?  If so, we should explicitly sort it here and then possibly make the code depend on the order.  The basic problem is that it assumes the first match is the best (`matching[0]`).  But if the sort order turns out the other way around, we might just need to take the last matching entry.

tomasz1986 replied:
The sort order doesn't cut it, because there can be 3+ variants of the language, so if we only rely on the order, then it won't ever match things like `zh-HK` (as it's between `zh-CN` and `zh-TW`).

The current iteration is not the greatest with the filter function basically duplicated, however it does seem to cover all edge cases. I've tested all different variants of English, all variants of Chinese, other single languages, etc. Specifically with English, `en-GB` needs to match `en-GB`, while `en` and all other `en-*` variants need to match just `en` (which in reality is `en-US` but the GUI uses simply `en` for the name).

The general idea is to look for exact matches first, and only when not found, go through the hyphenated variants, choosing the first matched. If you've got any tips how to simplify or improve the code, they will be very welcome.

acolomb commented on gui/default/syncthing/core/localeService.js:
```diff
@@ -78,14 +79,16 @@ angular.module('syncthing.core')
-                                    return possibleLang.indexOf(lang) === 0;
-                                } else {
-                                    return lang.indexOf(possibleLang) === 0;
+                                // Match "en" with "en-US" but not "fi" with "fil",
+                                // and match "en-US" with "en" but not "fil" with "fi".
```
```suggestion
                                // Match "en" with "en-US" but not "fi" with "fil", and vice versa
```

tomasz1986 replied:
Done.

acolomb commented:
Until now, we did not fall back to a lesser specified variant automatically.  For example, if a browser is configured to prefer "en-US", then we should not fall back to "en" automatically, just as we don't fall back from on to another "zh-*" variant.  I'd expect browsers to either automatically specify the broader codes ("en") in addition, or the user to configure them next to the more specific variant.  This avoids wrong conclusions and allows an actual preference ordering.  We should not introduce this fallback direction IMHO.

At least that's how I understand the existing code comment.  `Find the first language in the list provided by the user's browser that is a prefix of a language we have available.`  The actual bug is that "prefix" here means strictly a substring, while the first token before a dash should always be considered in full.

I do have some ideas how to simplify the code, but first we need to agree on that goal.  Should our "en" language be matched by any "en-XX" variant that the browser sent?  I think it should not.

tomasz1986 commented:
> Until now, we did not fall back to a lesser specified variant automatically.

The current code is broken though, i.e. it does specifically that. The browser requests `en-GB`, but the GUI displays `en` (aka `en-US`).

> For example, if a browser is configured to prefer "en-US", then we should not fall back to "en" automatically, just as we don't fall back from one to another "zh-*" variant.

It already does fall back to `en` (as the default language). With `zh-*`, the situation is a bit different, but if there is no fall back, it will just display the UI in English.

> I do have some ideas how to simplify the code, but first we need to agree on that goal. Should our "en" language be matched by any "en-XX" variant that the browser sent? I think it should not.

Why not? This should happen only if the specific variant of `en-XX` doesn't exist. On the other hand, I can see the point why `zh-TW` should not fall back to `zh-CN` (as the current iteration of the PR does).

acolomb commented:
> The current code is broken though, i.e. it does specifically that. The browser requests `en-GB`, but the GUI displays `en` (aka `en-US`).

That should be fixed while we're at it, IMHO.

> **Edit**: Basically, what happens in real life is that the browser requests `ko`, but our UI is available in `ko-KR` only. I think it should fall back to `ko-KR` in this case.

Yes, this direction is fine.  The preference basically says "Korean, but I don't care about the exact variant", and we give it the one we have.

If the preference were `fr-CA`, which we don't have, it's questionable whether any other French variant should be preferred over the last-resort English fallback.

If the browser says the user prefers Canadian French (`fr-CA`) but also accepts any other French variation, then it's okay to serve `fr`or `fr-FR` if we don't have `fr-CA` specifically.  But that choice should be left to the user (browser configuration) whether the broader variant (and therefore any other specific variant as well) is acceptable.

I for example have configured English (US) and German in Firefox.  The resulting preference includes `en` in 2nd place, after `en-US`.  According to https://manytools.org/http-html-text/browser-language/

tomasz1986 commented:
Just to get things straight then:

1. Requested `en-AU` which is not available. `en` is available but do not fall back to it automatically.
2. Requested `ko` which is not available but `ko-KR` is available. Fall back to `ko-KR` automatically.

Is this direction correct? From less specific to more specific only?

> I for example have configured English (US) and German in Firefox. The resulting preference includes `en` in 2nd place automatically, after `en-US`.

Yeah, it does this for English specifically but not for other languages. For example, if you select Chinese (Hong Kong), it requests just `zh-HK`. Does it also request German in Firefox when configured like that? In Chromium, only the first language from the list is requested.

tomasz1986 commented:
I've rewritten the whole thing. Now it only matches `en-GB` with `en-GB` exactly or `en` with `en-GB` (if no `en` found). It doesn't match `en-GB` with just `en`.

I've also changed some variable names, mainly `lang` to `browserLang`, and `matching` to `matchingLang`, as with the previous names, I was constantly confused which one was which… I hope this is fine.

acolomb reviewed (changes requested):
I think this is getting too complicated with the reverse sort order and all.  Let's take it back to a prototype and test concrete scenarios outside of the Syncthing GUI, then discuss any results that need changing.  Here's my experimentation code, tested on https://www.jsplayground.dev/

```javascript
function foo(langs) {
    // Find the first language in the list provided by the user's browser
    // that is a prefix of a language we have available. That is, "en"
    // sent by the browser will match "en" or "en-US", while "zh-TW" will
    // match only "zh-TW" and not "zh" or "zh-CN".

    var i,
        browserLang,
        matching,
        locale = _defaultLocale;  // Fallback if nothing matched

    for (i = 0; i < langs.length; i++) {
        browserLang = langs[i];
        if (browserLang.length < 2) {
            continue;
        }
        matching = _availableLocales.find(function (possibleLang) {
            // The langs returned by the /svc/langs call will be in lower
            // case. We compare to the lowercase version of the language
            // code we have as well.
            possibleLang = possibleLang.toLowerCase();
            if (possibleLang.indexOf(browserLang) !== 0) {
                // Prefix does not match
                return false;
            }
            console.log("compare", browserLang, possibleLang);
            if (possibleLang.length > browserLang.length) {
                // Must match up to the next hyphen separator
                return possibleLang[browserLang.length] === '-';
            }
            console.log("same length, exact match");
            // Same length, exact match
            return true;
        });

        if (matching) {
            console.log("matching", browserLang, matching);
            locale = matching;
            break;
        }
    }
    useLocale(locale);
}

var _availableLocales = ["ar","bg","ca","ca@valencia","cs","da","de","el","en","en-GB","es","eu","fil","fr","fy","ga","hi","hu","id","it","ja","ko-KR","lt","nl","pl","pt-BR","pt-PT","ro-RO","ru","sk","sl","sv","tr","uk","zh-CN","zh-HK","zh-TW"];
var _defaultLocale = "en";
function useLocale(locale) { console.log("Result:", locale, "\n") }
foo(["zh-TW", "ko", "en-gb!"]);
foo(["zh!", "ko", "en-us"]);
foo(["zh", "ko!", "en-us"]);
foo(["pt", "!en", "de"]);
```

I can make a separate PR with these changes if you like and find it functional in your testing.  Otherwise, please provide parameter sets that fail and what you think the result should be.  Of course, you can integrate that variant into this PR as well.

tomasz1986 commented on gui/default/syncthing/core/localeService.js:
```diff
@@ -59,34 +59,47 @@ angular.module('syncthing.core')
+                                possibleLang = guiLang.toLowerCase();
+                                if (possibleLang === browserLang) {
+                                    // Skip further checking if exact match found.
+                                    return matchingLang = guiLang;
```
Yeah, at this point it's just misused. It should be a simple loop.

Your code seems great, but I didn't consider `.find` due to worse browser support than `.filter`. Mainly, Android 5 with its stock WebView doesn't support it (and of course other older browsers).

What do you think about this version still using `.filter`?

```
function foo(langs) {
    // Find the first language in the list provided by the user's browser
    // that is a prefix of a language we have available. That is, "en"
    // sent by the browser will match "en" or "en-US", while "zh-TW" will
    // match only "zh-TW" and not "zh" or "zh-CN".

    var i,
        browserLang,
        matching,
        locale = _defaultLocale;  // Fallback if nothing matched

    for (i = 0; i < langs.length; i++) {
        browserLang = langs[i];
        if (browserLang.length < 2) {
            continue;
        }
        matching = _availableLocales.filter(function (possibleLang) {
            // The langs returned by the /svc/langs call will be in lower
            // case. We compare to the lowercase version of the language
            // code we have as well.
            possibleLang = possibleLang.toLowerCase();
            if (possibleLang.indexOf(browserLang) !== 0) {
                // Prefix does not match
                return false;
            }
            console.log("compare", browserLang, possibleLang);
            if (possibleLang.length > browserLang.length) {
                // Must match up to the next hyphen separator
                return possibleLang[browserLang.length] === '-';
            }
            console.log("same length, exact match");
            // Same length, exact match
            return true;
        });

        if (matching.length >= 1) {
            console.log("matching", browserLang, matching[0]);
            locale = matching[0];
            break;
        }
    }
    useLocale(locale);
}

var _availableLocales = ["ar","bg","ca","ca@valencia","cs","da","de","el","en","en-GB","es","eu","fil","fr","fy","ga","hi","hu","id","it","ja","ko-KR","lt","nl","pl","pt-BR","pt-PT","ro-RO","ru","sk","sl","sv","tr","uk","zh-CN","zh-HK","zh-TW"];
var _defaultLocale = "en";
function useLocale(locale) { console.log("Result:", locale, "\n") }
foo(["zh-tw", "ko", "en-gb!"]);
foo(["zh!", "ko", "en-us"]);
foo(["zh", "ko!", "en-us"]);
foo(["pt", "!en", "de"]);
```

I only changed

```
matching = _availableLocales.find(function (possibleLang) {
```

to

```
matching = _availableLocales.filter(function (possibleLang) {
```

and

```
if (matching) {
    console.log("matching", browserLang, matching);
    locale = matching;
    break;
}
```

to

```
if (matching.length >= 1) {
    console.log("matching", browserLang, matching[0]);
    locale = matching[0];
    break;
}
```

The results seem to be basically the same:

- using `.find`

```
compare zh-tw zh-tw 
same length, exact match 
matching zh-tw zh-TW 
Result: zh-TW 
 
compare ko ko-kr 
matching ko ko-KR 
Result: ko-KR 
 
compare zh zh-cn 
matching zh zh-CN 
Result: zh-CN 
 
compare pt pt-br 
matching pt pt-BR 
Result: pt-BR
```

- using `.filter`

```
compare zh-tw zh-tw 
same length, exact match 
matching zh-tw zh-TW 
Result: zh-TW 
 
compare ko ko-kr 
matching ko ko-KR 
Result: ko-KR 
 
compare zh zh-cn 
compare zh zh-hk 
compare zh zh-tw 
matching zh zh-CN 
Result: zh-CN 
 
compare pt pt-br 
compare pt pt-pt 
matching pt pt-BR 
Result: pt-BR
```

acolomb replied:
Wow, this compatibility stuff is getting really unnerving.  Can't wait to get rid of all this cumbersome crap ASAP.  MDN gives a green light "Baseline Widely available" rating for the function, but that seems to be still not enough.  We should pound some rocks together to make fire.

The latter point should be addressed in browsers.  I'm really opposed to doing more guesswork in this matter.  There are RFCs covering how it should be matched, and we can't accomodate every piece of broken software on the other side that doesn't implement them correctly.  And it's just the default value, any saved user choice will override it anyway, and that's what counts.

tomasz1986 replied:
> Wow, this compatibility stuff is getting really unnerving. Can't wait to get rid of all this cumbersome crap ASAP. MDN gives a green light "Baseline Widely available" rating for the function, but that seems to be still not enough. We should pound some rocks together to make fire.

Android will always lag behind. It's the nature of the beast 😉. Nowadays, Google does update WebView via the Play Store, but this doesn't help if the specific manufactured signed the built-in WebView app with their own signature (hence allowing new versions only via their own system updates, which for low-end devices may mean no updates). Also, the largest Android market, i.e. China, doesn't use Play Store at all (as it's banned there), so Google has no influence on the WebView updates there.

acolomb commented on gui/default/syncthing/core/localeService.js:
```diff
@@ -59,17 +59,17 @@ angular.module('syncthing.core')
-                            lang,
+                            browserLang,
                             matching,
                             locale = _defaultLocale;
```
```suggestion
                            locale = _defaultLocale;  // Fallback if nothing matched
```

acolomb commented on gui/default/syncthing/core/localeService.js:
```diff
@@ -78,11 +78,16 @@ angular.module('syncthing.core')
-                                } else {
-                                    return lang.indexOf(possibleLang) === 0;
+                                if (possibleLang.indexOf(browserLang) !== 0) {
+                                    // Prefix does not match.
```
```suggestion
                                    // Prefix does not match
```
Please let's not get started on this.  It's just code comments and for such short notes, we really don't need to pretend it's part of a prose text.  Please stick to the established style in this point.

tomasz1986 replied:
All fixed, although to be honest, there is no established style (especially if you go through all the JS files, not just this one). I think it all depends on who at what time wrote the specific code and comments.

It would be nice to have a specific style for HTML and JS files defined in the Docs though, which everyone would be expected to follow, as currently it really is all over the place…

acolomb replied:
Yes, something similar to `gofmt` would be helpful.  But before we start putting effort into such a thing, we should try to modernize and get to a more future-proof framework basis.  AngularJS in the currently used version is pretty outdated and our use of it rather monolithic, where individual components would help to better structure it.

tomasz1986 replied:
Yeah, but moving the GUI to another framework would basically mean a fulltime job, so I highly doubt something like that will happen in the near future, if ever. It may be easier to rewrite the whole thing from scratch 😅 (which will also likely require a lot of time).

I mean very basic things like 1) Don't use full stops in single-line comments, 2) Limit comments to xx characters per line, 3) Indent the code with x spaces / tabs, 4) Use HTML5 only, etc. Basically something like https://google.github.io/styleguide/jsguide.html#formatting. Maybe simply choose an established style guide for each language and link to it?

One advantage of the current Web GUI code is that it's been tested thoroughly in various environments. It's old-school but at same time stable, and the core doesn't change at all, so you know what to expect.

acolomb replied:
> Anyway, this part of the discussion probably belongs to the forum, not here 🙂.

Absolutely :-)

I've quickly read over the referenced Google Style guidelines.  They sound sensible AFAICS.  Very elaborate, as in many things to remember if you're just an occasional JS programmer.  But specifically the topic of the full stop in (inline code) comments is not clarified, only for JSDoc strings.  I stand by my opinion that it's just superfluous because these inline comments (in contrast to API docs) are never looked at outside of the code's context.

I would welcome an official recommendation on which style guide to follow, but that should be discussed on the forum, and also where to document it.

acolomb commented on gui/default/syncthing/core/localeService.js:
```diff
@@ -78,11 +78,16 @@ angular.module('syncthing.core')
                                 // case. We compare to the lowercase version of the language
```
The `/rest/langs` in the comment above, I had changed that on purpose.  It was renamed at some point and now lives at `/svc/lang`.

*GitHub wouldn't let me add a suggestion to that line though.*

tomasz1986 replied:
Done.

acolomb commented on gui/default/syncthing/core/localeService.js:
```diff
@@ -78,11 +78,16 @@ angular.module('syncthing.core')
+                                    return false;
                                 }
+                                if (possibleLang.length > browserLang.length) {
+                                    // Must match up to the next hyphen separator.
```
```suggestion
                                    // Must match up to the next hyphen separator
```

acolomb commented on gui/default/syncthing/core/localeService.js:
```diff
@@ -78,11 +78,16 @@ angular.module('syncthing.core')
+                                    // Must match up to the next hyphen separator.
+                                    return possibleLang[browserLang.length] === '-';
+                                }
+                                // Same length, exact match.
```
```suggestion
                                // Same length, exact match
```

acolomb reviewed (changes requested):
Well I'm not super happy with using `.filter()` because it seems enough to abort on first match.  But in this case at least it's easier to see all possible matches and validate the algorithm works as intended.

Some comment fixes, see below.
